torch/csrc/api/include/torch/nn/modules/any.h

```diff
@@ -0,0 +1,347 @@
+#pragma once
+
+#include <torch/detail/static.h>
+#include <torch/nn/module.h>
+
+#include <torch/csrc/autograd/variable.h>
+#include <torch/csrc/utils/memory.h>
+#include <torch/csrc/utils/variadic.h>
+
+#include <memory>
+#include <type_traits>
+#include <typeinfo>
+#include <utility>
+#include <vector>
+
+namespace torch {
+namespace nn {
+
+/// A class to store a type erasd module, whose `forward()` method can be
+/// invoked, with dynamic type checking.
+class AnyModule {
+ public:
+  /// A type-erased value.
+  class Value;
+
+  /// Constructs an `AnyModule` from a `shared_ptr` to concrete module object.
+  template <typename ModuleType>
+  explicit AnyModule(std::shared_ptr<ModuleType> module);
```

Comment:
Can we get this for operator= as well? Because in module construction, sometimes it's nice to use AnyModule in a Module, which means it's predeclared somewhere and I need to assign it to another module.

Reply:
yes I'll add it

Reply:
Actually this just works naturally in C++, since the right hand side of an `=` expression will construct a new `AnyModule`, which is then move-assigned:
```cpp
struct M : nn::Module {
      explicit M(int value_) : value(value_) {}
      int value;
      int forward(float x) {
        return x;
      }
    };
AnyModule any;
REQUIRE(any.is_empty());
any = AnyModule(M{5});
REQUIRE(!any.is_empty());
REQUIRE(any.get<M>().value == 5);
```

Reply:
But I've added an empty state

Reply:
Sorry, I meant something like

```
any = M{5};
```